Update document for Creating New Generators section

wenjie-p opened the pull request:
This is my first PR to fix the comment from https://github.com/JuliaLang/julia/issues/33008

wenjie-p committed:
Update document for Creating New Generators section

This is my first PR to fix the comment from https://github.com/JuliaLang/julia/issues/33008

wenjie-p committed:
Remove Trailing Space 

Trailing space on line 286 was removed.

rfourquet commented on stdlib/Random/docs/src/index.md:
```diff
@@ -283,8 +283,10 @@ Note: `Sampler(rng, x)` is simply a shorthand for `Sampler(rng, x, Val(Inf))`, a
+2) make sure that `copy` is possible for the new generator, otherwise, (i.e. hardware-based RNG ),
+   an error should be thrown;
+3) other documented `rand` methods accepting an `AbstractRNG` should work out of the box,
+   (provided the methods from 1) and 2) what are relied on are implemented),
```
Here point 2) is not relevant for whether `rand` works. I suggest that you move your point 2 at the end of the list, i.e. as point 3, as for the other 2 existing points, it's the API of the `rand` function specifically which is discussed.

rfourquet commented on stdlib/Random/docs/src/index.md:
```diff
@@ -283,8 +283,10 @@ Note: `Sampler(rng, x)` is simply a shorthand for `Sampler(rng, x, Val(Inf))`, a
-2) other documented `rand` methods accepting an `AbstractRNG` should work out of the box,
-   (provided the methods from 1) what are relied on are implemented),
+2) make sure that `copy` is possible for the new generator, otherwise, (i.e. hardware-based RNG ),
+   an error should be thrown;
```
I suggest instead: "make sure that `copy` is implemented for the new generator when possible, otherwise (e.g. hardware-based RNG), an error should be thrown."

But I'm not convinced that a specific error should be thrown, I believe that the default `MethodError` is enough.
So another way to put it is simply: "`copy` for RNGs should be implemented when possible."
I will let @tpapp comment on whether he thinks more information should be provided.

yuebanyishenqiu replied:
Thanks, I got it. Please let me know if there is something I can do for this.

wenjie-p committed:
Update document for Create New Generators section

There are two modifications in this commit:
a) Point 2) was adjusted to Point 3);
b) The space between RNG and the closing parentheses was removed.

wenjie-p committed:
Update document for Creating New Generator section

I have updated the content, sorry for that!

tpapp commented on stdlib/Random/docs/src/index.md:
```diff
@@ -285,7 +285,8 @@ The API is not clearly defined yet, but as a rule of thumb:
    (provided the methods from 1) what are relied on are implemented),
-   but can of course be specialized for this RNG if there is room for optimization.
+   but can of course be specialized for this RNG if there is room for optimization;
+3) `copy` for RNGs should be implemented when possible.
```
```suggestion
3) `copy` for pseudo-RNGs should return an independent copy that generates the exact same random sequence as the original from that point when called in the same way. When this is not feasible (eg hardware-based RNGs), `copy` does not need to be implemented.
```

wenjie-p committed:
Update stdlib/Random/docs/src/index.md

Update doc for Creating New Generator section

Co-Authored-By: Tamas K. Papp <tkpapp@gmail.com>

wenjie-p committed:
Update index.md

Reformatting the content in index.md

wenjie-p committed:
Update index.md

Replace tab with white space

tpapp commented:
@yuebanyishenqiu: thanks for doing this. As far as I am concerned this fixes #33008 sufficiently for the time being.

rfourquet commented on stdlib/Random/docs/src/index.md:
```diff
@@ -285,7 +285,10 @@ The API is not clearly defined yet, but as a rule of thumb:
+   but can of course be specialized for this RNG if there is room for optimization;
+3) `copy` for pseudo-RNGs should return an independent copy that generates the exact same random sequence as the
+   original from that point when called in the same way. When this is not feasible (eg hardware-based RNGs),
+   `copy` does not need to be implemented.
```
Sorry to nitpick, but the last sentence sounds superfluous to me: it amouts to say "if it can't be done sensibly, you don't need to do it", it's kinda obvious. But if we insist on clarifying that point, I believe we should say "`copy` _must_ not be implemented".

tpapp replied:
Good point.

yuebanyishenqiu replied:
Thanks to all of you! Point 3 has been modified according to the review.

wenjie-p committed:
Update index.md

Point 3 has been modified

rfourquet commented on stdlib/Random/docs/src/index.md:
```diff
@@ -285,7 +285,10 @@ The API is not clearly defined yet, but as a rule of thumb:
-   but can of course be specialized for this RNG if there is room for optimization.
+   but can of course be specialized for this RNG if there is room for optimization;
+3) `copy` for pseudo-RNGs should return an independent copy that generates the exact same random sequence as the
+   original from that point when called in the same way. When this is not feasible (eg hardware-based RNGs),
```
Isn't the norm to spell "e.g." with 2 dots?

yuebanyishenqiu replied:
Sure! Will update the content.

wenjie-p committed:
Update index.md
